llama.cpp

```diff
@@ -10003,143 +10024,82 @@ struct llm_tokenizer_bpe {
+    std::vector<size_t> regex_preprocess(const std::wstring & text, const std::vector<size_t> & offsets, const std::wstring & regex_expr) {
+        std::wregex expr(regex_expr);
+        std::vector<size_t> bpe_words; // stroe the offset of each word
+        bpe_words.reserve(offsets.size()); // Reserve memory for the approximate size
+        size_t start = 0;
+        for (auto offset : offsets) {
+            std::wcregex_iterator it(text.data() + start, text.data() + start + offset, expr);
+            std::wcregex_iterator end;
 
-            if (split_condition) {
-                if (token.size()) {
-                    bpe_words.emplace_back(token);
+            int64_t start_idx = 0;
+            while (it != end) {
+                std::wcmatch match = *it;
+                if (match.position() > start_idx) {
+                    bpe_words.emplace_back(match.position() - start_idx);
                 }
-                token = utf_char;
-                collecting = false;
-                collecting_letter = false;
-                collecting_numeric = false;
-                collecting_special = false;
-                collecting_whitespace_lookahead = false;
+                bpe_words.emplace_back(match.length());
+                start_idx = match.position() + match.length();
+                ++it;
             }
-            else {
-                token += utf_char;
+
+            if (start_idx < (int64_t) offset) {
+                bpe_words.emplace_back(offset - start_idx);
             }
+            start += offset;
         }
 
-        for (std::string & word : bpe_words) {
-            std::string encoded_token = "";
-            for (char & c : word) {
-                encoded_token += unicode_byte_to_utf8(c);
-            }
-            bpe_encoded_words.emplace_back(encoded_token);
+        return bpe_words;
+    }
+
+    std::vector<std::string> regex_bpe_preprocess(const std::string & text, const std::vector<std::wstring> & regex_exprs) {
+        std::wstring wtext = from_utf8(text);
+
+        std::vector<size_t> bpe_offsets = {wtext.size()};
+
+        for(auto & regex_expr : regex_exprs) {
+            bpe_offsets = regex_preprocess(wtext, bpe_offsets, regex_expr);
         }
 
-        return bpe_encoded_words;
+        std::vector<std::string> bpe_words;
+        bpe_words.reserve(bpe_offsets.size()); // Reserve memory for the approximate size
+        size_t start = 0;
+        for(size_t & offset : bpe_offsets){
+            bpe_words.emplace_back(to_utf8(std::wstring(wtext, start, offset)));
+            start += offset;
+        }
+
+        return byte_encoding_process(bpe_words);
+    }
+
+    std::vector<std::string> bpe_gpt2_preprocess(const std::string & text) {
+        return regex_bpe_preprocess(text, get_gpt2_regex());
+    }
+
+    std::vector<std::string> bpe_deepseek_coder_preprocess(const std::string & text) {
+        return regex_bpe_preprocess(text, get_deepseek_coder_regex());
+    }
```

Comment:
Following my previous comment, this should eventually become:

```suggestion
    std::vector<std::string> bpe_deepseek_coder_preprocess(const std::string & text) {
        return unicode_regex_split(text, {
	        "[\\p{P}\\$\\+<=>\\^~\\|]+",
	        "'s|'t|'re|'ve|'m|'ll|'d| ?\\p{L}+| ?\\p{N}+| ?[^\\s\\p{L}\\p{N}]+|\\s+(?!\\S)",
	        "[0-9][0-9][0-9]"
	        "\\s?\\p{L}+",
	        "\\s?\\p{P}+",
	        "\\p{N}",
        });
    }
```